### PR TITLE
Add support for custom adjacency context in masked autoregressive transform 

### PR DESCRIPTION
**Pull request by aalmodovares**
This PR implements [#64](https://github.com/probabilists/zuko/issues/64), adding support for user-defined structural constraints in the context features of MaskedAutoregressiveTransform.

Allows specifying an adjacency matrix of shape (features, features + context).

The right-most columns correspond to context-related constraints.

Adds assertions to verify shape validity.

Includes a test to ensure correct behavior.

No additional assumptions are made about the structure of the context adjacency beyond its shape.

Feel free to suggest any changes or additional constraints to enforce.

### CODE REVIEW

File: zuko/flows/autoregressive.py [hunk 1]
```diff
@@ -120,21 +122,31 @@ def __init__(
...
             assert adjacency.ndim == 2, "'adjacency' should be a matrix."
             assert adjacency.shape[0] == features, f"'adjacency' should have {features} rows."
-            assert adjacency.shape[1] == features, f"'adjacency' should have {features} columns."
+            assert adjacency.shape[1] == features or adjacency.shape[1] == features + context, (
+                f"'adjacency' should have {features} or {features + context} columns."
+            )
+
+            has_context = adjacency.shape[1] == features + context
+            context_adjacency = adjacency[:, features:] if has_context else None
+            adjacency = adjacency[:, :features]
```
Comment (francois-rozet) [inline zuko/flows/autoregressive.py:137]: I prefer a simple if-else than a ternary operator. 

```suggestion
            adjacency_context = adjacency[:, features:] if adjacency.shape[1] > features else None
            adjacency = adjacency[:, :features]
```

File: tests/test_flows.py [hunk 2]
```diff
@@ -153,3 +153,41 @@ def test_adjacency_matrix():
...
+    # With a valid adjacency matrix
+    order = torch.randperm(5)
+    adjacency = torch.rand((5, 5)) < 0.25
+    adjacency = adjacency + torch.eye(5, dtype=bool)
+    adjacency = torch.tril(adjacency)
+    adjacency = adjacency[order, :][:, order]
+
+    # context adjacency
+    adjacency_context = torch.rand((5, 2)) < 0.25
+    valid_adjacency = torch.concat((adjacency, adjacency_context), dim=1)
```
Comment (francois-rozet) [inline tests/test_flows.py:170]:
```suggestion
    adjacency_valid = torch.cat((adjacency, adjacency_context), dim=1)
```

File: tests/test_flows.py [hunk 3]
```diff
@@ -153,3 +153,41 @@ def test_adjacency_matrix():
...
+
+    t = T(features=5, context=2, adjacency=valid_adjacency)
```
Comment (francois-rozet) [inline tests/test_flows.py:172]:
```suggestion
    t = T(features=5, context=2, adjacency=adjacency_valid)
```

File: tests/test_flows.py [hunk 4]
```diff
@@ -153,3 +153,41 @@ def test_adjacency_matrix():
...
+    ladj = torch.linalg.slogdet(J).logabsdet
+
+    assert torch.allclose(t(c).log_abs_det_jacobian(x, y), ladj, atol=1e-4)
+    assert torch.allclose(J.diag().abs().log().sum(), ladj, atol=1e-4)
+
+    invalid_adjacency_context = torch.rand((5, 1)) < 0.25
+    invalid_adjacency = torch.concat((adjacency, invalid_adjacency_context), dim=1)
+
+    with pytest.raises(AssertionError, match="'adjacency' should have 5 or 7 columns."):
+        t = T(features=5, context=2, adjacency=invalid_adjacency)
```
Comment (francois-rozet) [inline tests/test_flows.py:193]:
```suggestion
    adjacency_invalid_context = torch.rand((5, 1)) < 0.25
    adjacency_invalid = torch.cat((adjacency, adjacency_invalid_context), dim=1)

    with pytest.raises(AssertionError, match="'adjacency' should have 5 or 7 columns."):
        t = T(features=5, context=2, adjacency=adjacency_invalid)
```

File: tests/test_flows.py [hunk 5]
```diff
@@ -153,3 +153,41 @@ def test_adjacency_matrix():
...
+
+    x, c = randn(5), randn(2)
+    y = t(c)(x)
+
+    J = torch.autograd.functional.jacobian(t(c), x)
+    assert (J[~adjacency] == 0).all()
+
+    assert y.shape == x.shape
+    assert y.requires_grad
+    assert torch.allclose(t(c).inv(y), x, atol=1e-4)
```
Comment (francois-rozet) [inline tests/test_flows.py:182]:
```suggestion
    assert y.shape == x.shape
    assert y.requires_grad
    assert torch.allclose(t(c).inv(y), x, atol=1e-4)
    
    J = torch.autograd.functional.jacobian(t(c), x)

    assert (J[~adjacency] == 0).all()
```

File: zuko/flows/autoregressive.py [hunk 6]
```diff
@@ -40,7 +40,9 @@ class MaskedAutoregressiveTransform(LazyTransform):
         order: A feature ordering. If :py:`None`, use :py:`range(features)` instead.
         adjacency: An adjacency matrix describing the transformation graph. If
             `adjacency` is provided, `order` is ignored and `passes` is replaced by the
-            diameter of the graph.
+            diameter of the graph. Its shape must be `(features, features)`
```
Comment (francois-rozet) [inline zuko/flows/autoregressive.py:43]:
```suggestion
            diameter of the graph. Its shape must be either `(features, features)`
```

File: zuko/flows/autoregressive.py [hunk 7]
```diff
@@ -120,21 +122,31 @@ def __init__(
             self.order = torch.div(order, ceil(features / self.passes), rounding_mode="floor")
 
             adjacency = self.order[:, None] > self.order
+            context_adjacency = None
```
Comment (francois-rozet) [inline zuko/flows/autoregressive.py:125]:
```suggestion
            adjacency_context = None
```

File: zuko/flows/autoregressive.py [hunk 8]
```diff
@@ -120,21 +122,31 @@ def __init__(
...
+
             assert adjacency.diag().all(), "'adjacency' should have ones on the diagonal."
 
             adjacency = adjacency * ~torch.eye(features, dtype=bool)
 
             self.passes = self._dag_diameter(adjacency)
 
         if context > 0:
+            if context_adjacency is None:
+                context_adjacency = torch.ones((features, context), dtype=bool)
```
Comment (francois-rozet) [inline zuko/flows/autoregressive.py:147]:
```suggestion
            if adjacency_context is None:
                adjacency_context = torch.ones((features, context), dtype=bool)
```

File: zuko/flows/autoregressive.py [hunk 9]
```diff
@@ -120,21 +122,31 @@ def __init__(
...
             adjacency = torch.cat(
-                (adjacency, torch.ones((features, context), dtype=bool)),
+                (adjacency, context_adjacency),
```
Comment (francois-rozet) [inline zuko/flows/autoregressive.py:149]:
```suggestion
                (adjacency, adjacency_context),
```

Could you also make this operation a single line?

**Review by francois-rozet — COMMENTED** [review]
Hi @aalmodovares, I reviewed the proposed changes. These look very appropriate, and the tests as well. Good job! I left a few comments, mainly regarding variable names and a bit of formatting. Tell me if you agree with the suggested changes. I will merge right after. Thanks a lot!

**Comment by aalmodovares** [comment]
Hi @francois-rozet. Thank you for accepting my PR! 

I agree with all your suggestions, I have implemented them already and checked everything works. All the suggestions are solved.

Thank you also for the brevity. Feel free to reach me again if something can be improved! :)

File: zuko/flows/autoregressive.py (hunk 9 above)
Comment (aalmodovares) [reply zuko/flows/autoregressive.py]: Sure!

File: zuko/flows/autoregressive.py [hunk 10]
```diff
@@ -120,23 +122,29 @@ def __init__(
...
             self.passes = self._dag_diameter(adjacency)
 
         if context > 0:
-            adjacency = torch.cat(
-                (adjacency, torch.ones((features, context), dtype=bool)),
-                dim=1,
-            )
+            if adjacency_context is None:
+                adjacency_context = torch.ones((features, context), dtype=bool)
+            adjacency = torch.cat((adjacency, adjacency_context),dim=1)
```
Comment (francois-rozet) [inline zuko/flows/autoregressive.py:147]:
```suggestion
            adjacency = torch.cat((adjacency, adjacency_context), dim=1)
```

File: tests/test_flows.py [hunk 11]
```diff
@@ -153,3 +153,41 @@ def test_adjacency_matrix():
...
+    y = t(c)(x)
+
+    assert y.shape == x.shape
+    assert y.requires_grad
+    assert torch.allclose(t(c).inv(y), x, atol=1e-4)
+
+    J = torch.autograd.functional.jacobian(t(c), x)
+
+    assert (J[~adjacency] == 0).all()
+    ladj = torch.linalg.slogdet(J).logabsdet
```
Comment (francois-rozet) [inline tests/test_flows.py:184]:
```suggestion
    assert (J[~adjacency] == 0).all()

    ladj = torch.linalg.slogdet(J).logabsdet
```

**Comment by francois-rozet** [comment]
The PR has been merged 🥳 Very good job on your first contribution to Zuko!

**Comment by aalmodovares** [comment]
Thank you!! I would be glad to collaborate again!
